src/SmoFacade/Listener.cs

```diff
@@ -130,19 +131,81 @@ private static string AgListenerName(string dataSource)
       return dotIndex >= 0 ? dataSource.Remove(dotIndex) : dataSource;
     }
 
-    private static Server AgListenerNameToServer(ref SqlConnectionStringBuilder connBuilder, string agInstanceName,
+    private static Server AgInstanceNameToServer(ref SqlConnectionStringBuilder connBuilder, string agInstanceName,
       string credentialName)
     {
-      var parts = agInstanceName.Split('\\');
-      if(parts.Length == 1)
-        connBuilder.DataSource = Dns.GetHostEntry(agInstanceName).HostName;
-      else if(parts.Length == 2)
-        // NamedInstances: chop instance name, resolve DNS, slap instance name back on!
-        connBuilder.DataSource = $"{Dns.GetHostEntry(parts[0]).HostName}\\{parts[1]}";
-      else
-        throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS");
-
-      return new Server(connBuilder.ToString(), credentialName);
+      try 
+      {
+        connBuilder.DataSource = ResolveDnsHostNameForInstance(agInstanceName, connBuilder.DataSource);
+        return new Server(connBuilder.ToString(), credentialName);
+      }
+      catch (Exception e)
+      {
+        throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS", e);
+      }
     }
+
+    /// <summary>
+    ///   Resolves 'agReplicaInstanceName' to a FQDN
+    ///   However on Unix OS, when 'val' in 'Dns.GetHostEntry(val)' is not a complete domain (i.e is just "abc", instead of "abc.def.com"), it fails intermittently
+    ///   Therefore, if dns lookup on just the instance name fails, we retry after appending the domain fragments from the listener to the instance name
+    /// </summary>
+    /// <param name="agReplicaInstanceName"> The name for an instance within the AG (for which we are trying to get the FQDN)</param>
+    /// <param name="agListenerDomain"> The complete domain for the AG listener</param>
+    private static string ResolveDnsHostNameForInstance(string agReplicaInstanceName, string agListenerDomain)
+    {
+      // Sometimes instances and listeners have ports or "named instances" (NI)
+      // Therefore, we strip them off before calling DNS.GetHostEntry() and then add them back to the result 
+      var (listenerDomain, listenerPortOrNi) = SplitDomainAndPort(agListenerDomain);
+      var (instanceName, instancePortOrNi) = SplitDomainAndPort(agReplicaInstanceName);

```

Comment:
I'm not big on acronyms in variable names... `Ni` --> `NamedInstance` is verbose but more readable

Reply:
I was debating this myself, it looked very long - so thought it might be better to name it `Ni` and put in a comment above
`// Sometimes instances and listeners have ports or "named instances" (NI)`

But yes, I can see your point and I now agree too - should go for more easy to read names